### PR TITLE
UnionExec array and nested array support

### LINKED ISSUE #1459: [FEA] Support union for nested types
**Is your feature request related to a problem? Please describe.**
Union only works for non-nested types right now. It would be nice to support it for nested types as well. This is mostly blocked on scalar support for nested types. When we do get around to implementing it we should be sure to test with [SPARK-32376](https://issues.apache.org/jira/browse/SPARK-32376) where a struct can have null columns added to it to make the two sides match. I am not sure if it will require any code changes on our part though.

### PR DESCRIPTION
**Pull request by rwlee**
Resolves #1459 

Depends on https://github.com/rapidsai/cudf/pull/9130 for unionByName list of struct support. Approved but not merged yet.

Adds type support for arrays and nested arrays for `union`, `unionAll`, `unionByName`. Also includes better UnionExec testing for nested types.

### CODE REVIEW

**Review by razajafri — APPROVED** [review]

**Comment by razajafri** [comment]
build

**Comment by jlowe** [comment]
Curious, this depends on a cudf feature that hasn't been merged but the tests pass.  This makes me think the tests aren't sufficiently exercising the code, otherwise a test should have failed without the requisite cudf feature.

**Comment by ttnghia** [comment]
> Curious, this depends on a cudf feature that hasn't been merged but the tests pass. This makes me think the tests aren't sufficiently exercising the code, otherwise a test should have failed without the requisite cudf feature.

This, this looks dangerous to me too :smiley:

**Comment by jlowe** [comment]
Converting this to a draft so this doesn't get accidentally merged by a passerby until this mystery is solved.

**Comment by rwlee** [comment]
build

**Comment by rwlee** [comment]
I turn this test error on and off with a very clear change within cudf, but for whatever reason the followup build still passed. The followup build was started before the PR was merged, but it was pretty close timing. If the blossom CI was delayed pulling the snapshot it may have picked up a cudf snapshot with the fix in it. Looking into it tomorrow, but with locally tests demonstrating cause and effect I think this is good to reopen/merge...

**Comment by rwlee** [comment]
Confirmed while testing with different versions of the snapshot jars, the new test correctly hits the dependent interleave column functionality.

**Review by razajafri — APPROVED** [review]
Approve the plugin side changes assuming it has been verified that this patch fails with older versions of cudf

**Comment by rwlee** [comment]
> Approve the plugin side changes assuming it has been verified that this patch fails with older versions of cudf

Verified offline with different snapshot builds. Fails with `cudf-21.10.0-20210902.122036-35-cuda11.jar` and `cudf-21.10.0-20210902.063732-34-cuda11.jar`, but passes with `cudf-21.10.0-20210903.122002-36-cuda11.jar`

**Review by revans2 — APPROVED** [review]

**Pull request merged**
